### ISSUE TITLE
ANF and Effect Handlers

### ISSUE THREAD

**Issue by cyberglot** [anoma/juvix#709]
tl;dr

**Goal** implement an ANF IR for named delimited continuations

**Main Integration** Witch, and later on with the whole codebase

**Gains** 
 - Re-work reduction
 - Optimisation opportunities
 - Easier-to-deal IR
 - Easier to translate to low-level languages

**Concerns** ~~Proof-caring translation is possible? How careful do we need to be?~~ Assume it is safe, we can verify later.

----

As an optimisation step, we want to have our core language be translated into ANF (A-normal form -- "A" may stand for "Administrative"). However, Algebraic Effects and Handlers are usually implemented as CPS-fied delimited continuations. To avoid multiple translations and unecessary rework, we have decided to implement Effect Handlers in ANF directly.

## ANF Overview

ANF is a common intermediate representation used in compilers, with a simple motif: pure programs stay untouched; effectful computations get sequentialised, e.g. everything that:
  - is guaranteed to terminate;
  - causes no side effects;
  - causes no control effects; and
  - never produces an error.

As an examaple, take the following expression  

```haskell
f (g x) (h y)
```

that is translated into ANF as

```haskell
let v0 = g x
    v1 = h y
in f v0 v1
```

## Implications for Juvix

The main reason to use ANF rather than CPS is how it makes the language easier to deal with. We will have to deal with Juvix code a lot, and reading CPS-fied code is a nightmare. Other reasons are that we can still apply  many code optimisations (e.g., change the order of many `let`s if necessary, while swapping continuations is a much harder task) and it translates well into stack machines and ASM-based machines (since the data-flow is more apparent).

The only non-obvious question is whether it is safe to have ANF-based proof transformations. It is unclear whether the process does or can carry proofs, or if the original semantics that have been proved carry over the transformation. After looking more into the issue, it seems that all production-level languages don't formally check if semantics/types were preserved after transformation. The solution for now amounts to informal/pen-and-paper reasoning of the correctness of transformation and tests. Later on, we can formally verify it with some known techniques (see References).

### ANF transformation and optimisations
- [The Essence of Compiling with Continuations](https://slang.soe.ucsc.edu/cormac/papers/pldi93.pdf) (Paper that introduces ANF)
- [Reasoning about Programs in Continuation-Passing Style](https://legacy.cs.indiana.edu/~sabry/papers/reasoning-lasc.ps)
- [A Functional Perspective on SSA Optimisation Algorithms](https://www.jantar.org/papers/chakravarty03perspective.pdf)

### Proof-carrying code and type perservation across transformations

- [Foundational Proof-Carrying Code](https://www.cs.princeton.edu/~appel/papers/fpcc.pdf)
- [PROOF-CARRYING CODE. DESIGN AND IMPLEMENTATION.](http://people.eecs.berkeley.edu/~necula/Papers/marktoberdorf.pdf)
- [Automatically Proving the Correctness of Compiler Optimizations](ftp://ftp.cs.washington.edu/homes/chambers/pldi03.pdf)
- [On the Correctness of Transformations in Compiler Back-Ends](https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.1082.9343&rep=rep1&type=pdf)
- [A structured approach to proving compiler optimizations based on dataflow analysis](https://xavierleroy.org/publi/proofs_dataflow_optimizations.pdf)

**Comment by cwgoes** [comment]
Is there any prior literature on ANF-based proof transformations?

**Comment by cyberglot** [comment]
> Thanks for the summary. Is there any prior literature on ANF-based proof transformations?

Added to the main issue.

**Comment by cwgoes** [comment]
I think this is fine to assume ANF transform is safe for now - if these forms are not equal we're doing something else very very wrong.

**Comment by cwgoes** [comment]
Part of idealised pipeline discussions.
